cmake/Gems.cmake

```diff
@@ -80,6 +80,72 @@ macro(o3de_gem_setup default_gem_name)
     o3de_pal_dir(pal_dir ${CMAKE_CURRENT_SOURCE_DIR}/Platform/${PAL_PLATFORM_NAME} "${gem_restricted_path}" "${gem_path}" "${gem_parent_relative_path}")
 endmacro()
 
+# resolve_all_dependent_gem_names_recursive
+#
+# Inspect a gem's dependencies recursively to collect of the list of gem names that the gem depends on
+#
+# \arg:gem_name(STRING) - Gem name whose "dependencies" will be queried from its gem.json
+# \arg:input_resolved_gem_names(LIST) - The current list of resolved gem name dependencies
+# \arg:output_resolved_gem_names(LIST) - The updated list of resolved gem name dependencies
+function(resolve_all_dependent_gem_names_recursive gem_name input_resolved_gem_names output_resolved_gem_names)
+
+    # Stop if the gem dependency has already been resolved
+    list(FIND input_resolved_gem_names ${gem_name} found_gem_name_index)
+    if (found_gem_name_index GREATER_EQUAL 0)
+        # Recursive stop condition
+        return()
+    endif()
+
+    # Attempt to read the dependencies for the gem
+    get_property(gem_path GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
+    if (gem_path)
+        o3de_read_json_array(gem_dependencies "${gem_path}/gem.json" "dependencies")
+    else()
+        return()
+    endif()
+
+    set(resolved_gem_names ${input_resolved_gem_names})
+
+    # Read in the dependencies recursively
+    foreach(enabled_gem_dependent_gem_name IN LISTS gem_dependencies)
+        list(APPEND resolved_gem_names "${enabled_gem_dependent_gem_name}")
+        resolve_all_dependent_gem_names_recursive(${enabled_gem_dependent_gem_name} ${resolved_gem_names} resolved_gem_names)
+    endforeach()
+
+    list(REMOVE_DUPLICATES resolved_gem_names)
+
+    set(${output_resolved_gem_names} ${resolved_gem_names} PARENT_SCOPE)
+
+endfunction()
+
+# resolve_all_dependent_gem_names
+#
+# Given a list of gem names, resolve all the dependent gems and build a list of additional
+# gems that need to be included for full dependency coverage
+#
+# \arg:input_gem_name_list(LIST) - The list of gem names to resolve all the dependencies
+# \arg:out_gem_name_list(LIST) - The list of additional gems that are not currently in the list to consider
+function(resolve_all_dependent_gem_names input_gem_name_list out_gem_name_list)
+
+    # Build up the list of all dependent gems (recursively)
+    unset(all_gems_name_list)
+    foreach(gem_name IN LISTS input_gem_name_list)
+        unset(dependent_gems_for_gem)
+        resolve_all_dependent_gem_names_recursive(${gem_name} "" dependent_gems_for_gem)
+        list(APPEND all_gems_name_list ${dependent_gems_for_gem})
+    endforeach()
+
+    # Remove duplicates and gem names that were already on the list
+    list(REMOVE_DUPLICATES all_gems_name_list)
+    foreach(gem_name IN LISTS input_gem_name_list)
+        list(REMOVE_ITEM all_gems_name_list ${gem_name})
+    endforeach()
+
+    list(SORT all_gems_name_list)
+    set(${out_gem_name_list} ${all_gems_name_list} PARENT_SCOPE)
```

Comment:
This doesn't perform any caching of the gem_names that already had it's dependencies resolved in a cmake GLOBAL property.

However we already gather the list of Gem dependencies for each Gem based on the active Gem "variants".
The [o3de_add_variant_dependencies_for_gem_dependencies](https://github.com/o3de/o3de/blob/8749168e1f79263395b945f1623b7314003bf366/cmake/Gems.cmake#L134-L141) should be sufficient for this purpose without the need to add a  "resolve_all_dependent_gem_names" function.

However if we are going to implement such a function, it should be done in a way that the result per Gem name is cache in a cmake GLOBAL property. Because  `ly_enable_gems` is called multiple times through out the CMake configuration step this function can gather the dependent gem names multiple times.

Two suggestions here:
1. The list of dependent gem names can be gathering in the `ly_enable_gems_delayed` function at the end of this file. That function is only called once at the end of the CMake configuration step and iterates over all  Gems.
2. The caching of the "Gem Name" -> "dependent Gem Names"  map can be done in the `Subdirectories.cmake` [reorder_dependent_gems_with_cycle_detection](https://github.com/o3de/o3de/blob/development/cmake/Subdirectories.cmake#L275-L282) function as it already recursively reads the "dependents" field for each Gem and [caches](https://github.com/o3de/o3de/blob/development/cmake/Subdirectories.cmake#L320-L321) the paths into the "Dependent:${gem_external_subdir_path}" GLOBAL property.
In that same function a mapping since we have access to that mapping of "Gem external subdirectory" -> "Gem dependencies external subdirectories", the `o3de_read_json_key` function can be just be used to read the "gem_name" field for each and then a new mapping GLOBAL property of "Gem name" -> "Gem dependencies names" can be added there.

Reply:
Thats a good suggestion, I can eliminate the additional loop and recursion if its already done somewhere (I didn't see that [reorder_dependent_gems_with_cycle_detection](https://github.com/o3de/o3de/blob/development/cmake/Subdirectories.cmake#L275-L282) function.

Reply:
Updated the PR, with some notes:
1. The gem dependency resolution on names needed to occur before the [LY_DELAYED_ENABLE_GEMS](https://github.com/spham-amzn/o3de/blob/79232a97d2bbe37457c0b6c20bb10fac7d9001c5/cmake/Gems.cmake#L439) property for the project, `ly_enable_gems_delayed` was too late in the process
2. The reason was that the [LY_DELAYED_LOAD_](https://github.com/spham-amzn/o3de/blob/79232a97d2bbe37457c0b6c20bb10fac7d9001c5/cmake/Gems.cmake#L557-L558) property was set from iterating on the results of `LY_DELAYED_ENABLE_GEMS_"${project}"`. 
3. Setting the global property allowed me to memo-ize the recursion and simplify the logic

Reply:
That is good with me.